### ISSUE TITLE
Wasm getting symbol collisions from libc

### ISSUE THREAD

**Issue by fengb** [ziglang/zig#5468]
_Note: I am not importing libc_

If I try to either import or export a symbol that's a part of libc, lld starts doing special things.

```zig
export fn log() {}
// generates a garbage body:
// (func $unreachable:log (type 0) (param f64) (result f64) unreachable)

export const sqrtf: u32 = 0;
// lld: error: symbol type mismatch: sqrtf
```

**Comment by kubkon** [comment]
I'm assuming you're compiling to `wasm32-freestanding`?

**Comment by fengb** [comment]
Yes, `zig build-lib broken.zig -target wasm32-freestanding --release-safe`

**Comment by kubkon** [comment]
OK, so we export several basic maths functions (including `sqrt` and `sqrtf`) plus memory primitive ops such as `memcopy`, `memcmp`, etc., in [std/special/c.zig]. I guess we could in principle only export memory ops when targeting `wasm32-freestanding` (skipping maths exports), but then, I think I'd like to keep `wasm32-wasi` free of those as well and that's currently not possible since we generate calls to those functions in the IR. I'm wondering, perhaps the correct solution would be to tweak the IR to search for mangled symbols such as `__sqrtf` instead of `sqrtf`? This way, we could still link them while not generating any collisions (or minimising the collision surface). Thoughts @fengb @andrewrk?

[std/special/c.zig]: https://github.com/ziglang/zig/blob/master/lib/std/special/c.zig

**Comment by pannous** [comment]
just wild guess out of nowhere : could it be related to log being a macro in math.h

**Comment by kubkon** [comment]
@fengb could you verify that this is no longer the case when targeting `wasm32-freestanding`? I believe it is still a problem when targeting `wasm32-wasi` though. For example, modifying @Vexu's test case

```zig
export const sqrtf: u32 = 2;
pub fn main() !void {
    try @import("std").testing.expect(sqrtf == 2);
}
```

we get wasm-ld multiple symbol def / type mismatch error:

```
>>> defined as WASM_SYMBOL_TYPE_DATA in zig-cache/o/463078301afe5226820887034da46d36/hmm.o
>>> defined as WASM_SYMBOL_TYPE_FUNCTION in /home/kubkon/.cache/zig/o/e7548511ae6d570164e23c830cb6f89c/libc.a(/home/kubkon/.cache/zig/o/e7548511ae6d570164e23c830cb6f89c/c.o)
```

**Comment by kubkon** [comment]
OK, so here's the tl;dr of the problem. If the Zig source contains any reference to a symbol from `libc.a(c.o)` where `c.o` is a translation unit containing the exports from `lib/std/special/c.zig`, then @fengb's collision problem and type mismatch still applies.

**Comment by kubkon** [comment]
After careful investigation, it looks like this is in part a limitation of `wasm-ld`. The idea for solving this that both Andrew and I arrived at was to have Zig's minimal `libc.a` contain only weak definitions. This way, if you exported a symbol in your project, it would immediately get strong linkage and override any definition from the `libc.a`. However, this is not that easy with `wasm-ld` since this linker seems to be doing some actual symbol type inference at link time, and this is specific to this linker only. One solution which will happen sooner or later is adding Wasm support to `zig ld` and have full control over considerations like this, but since this will be a bigger undertaking, I am pushing forward this issue until at least `0.9.0` release.

If anyone has other ideas, please share!

**Comment by r00ster91** [comment]
Interestingly, this works:
```
extern "custom_name" fn log() void;

export fn main() void {
    log();
}
```

**Comment by alexrp** [comment]
It's not clear to me that there's a bug here.

```
❯ zig4 build-lib -dynamic -target wasm32-freestanding repro.zig
error: wasm-ld:
error: wasm-ld: symbol type mismatch: sqrtf
    note: defined as WASM_SYMBOL_TYPE_DATA in /home/alexrp/.cache/zig/tmp/bb60f6a11c6985ca/repro_zcu.o
    note: defined as WASM_SYMBOL_TYPE_FUNCTION in /home/alexrp/.cache/zig/o/8f7e96121691fae55d5dc8e8c1577072/libcompiler_rt.a(/home/alexrp/.cache/zig/o/8f7e96121691fae55d5dc8e8c1577072/libcompiler_rt_zcu.o)
```

This just... makes sense? It is indeed a type mismatch with the compiler-rt symbol.

It works with `-fno-compiler-rt`.
